File: 2022/july/27-07-2022/flatten-binary-tree-to-linked-list.py
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        def traverse(node):
            if not node: return
            if node.left:
                next = node.left
                while next.right:
                    next = next.right
                next.right = node.right
                node.right = node.left
                node.left = None
            traverse(node.right)
        traverse(root)

File: 2022/july/27-07-2022/flatten-binary-tree-to-linked-list.rs
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
	pub fn flatten(root: &mut Option<Rc<RefCell<TreeNode>>>) {
		let mut root = root.clone();
		while let Some(r) = root {
			let mut r = r.borrow_mut();
			if let Some(mut prec) = r.left.clone() {
				loop {
					let right = prec.borrow().right.clone();
					if let Some(ri) = right {prec = ri;}
					else {break;}
				}
				let mut b = prec.borrow_mut();
				b.right = r.right.take();
				r.right = r.left.take()
			}
			root = r.right.clone();
		}
	}
}